File: nn/src/test/scala/com/intel/webscaleml/nn/nn/performTest/SpatialConvolutionPerform.lua
--
-- Created by yansh on 16 - 6 - 15.
--

require 'nn'

forwardIterations = 10
backwardIterations = 10
seed = 100

batchSize = 10


allTestCases = {
    -- AlexNet
    {IP = 3, OP = 64, kW = 11, kH = 11, dW = 4, dH = 4, padW = 2, padH = 2, iW = 224, iH = 224},
    {IP = 64, OP = 192, kW = 5, kH = 5, dW = 1, dH = 1, padW = 2, padH = 2, iW = 25, iH = 25},
    {IP = 191, OP = 384, kW = 3, kH = 3, dW = 1, dH = 1, padW = 1, padH = 1, iW = 12, iH = 12},
    {IP = 384, OP = 256, kW = 3, kH = 3, dW = 1, dH = 1, padW = 1, padH = 1, iW = 6, iH = 6},
    {IP = 256, OP = 256, kW = 3, kH = 3, dW = 1, dH = 1, padW = 1, padH = 1, iW = 3, iH = 3},
    --Cifar
    {IP = 3, OP = 64, kW = 3, kH = 3, dW = 1, dH = 1, padW = 1, padH = 1, iW = 224, iH = 224},
    {IP = 64, OP = 64, kW = 3, kH = 3, dW = 1, dH = 1, padW = 1, padH = 1, iW = 110, iH = 110},
    {IP = 64, OP = 128, kW = 3, kH = 3, dW = 1, dH = 1, padW = 1, padH = 1, iW = 54, iH = 54},
    {IP = 128, OP = 128, kW = 3, kH = 3, dW = 1, dH = 1, padW = 1 ,padH = 1 ,iW = 26, iH = 26},
    {IP = 128, OP = 256, kW = 3, kH = 3, dW = 1, dH = 1, padW = 1, padH = 1, iW = 13, iH = 13},
    {IP = 256, OP = 256, kW = 3, kH = 3, dW = 1, dH = 1, padW = 1, padH = 1, iW = 6, iH = 6},
    {IP = 256, OP = 512, kW = 3, kH = 3, dW = 1, dH = 1, padW = 1, padH = 1, iW = 3, iH = 3},
    {IP = 512, OP = 512, kW = 3, kH = 3, dW = 1, dH = 1, padW = 1, padH = 1, iW = 2, iH = 2},
    
    --GoogleNet
    {IP = 3, OP = 64, kW = 7, kH = 7, dW = 2, dH = 2, padW = 3, padH = 3, iW = 224, iH = 224},
    {IP = 64, OP = 64, kW = 1, kH = 1, dW = 1, dH = 1, padW = 0, padH = 0, iW = 54, iH = 54},
    {IP = 64, OP = 192, kW = 3, kH = 3, dW = 1, dH = 1, padW = 1, padH = 1, iW = 27, iH = 27},
    {IP = 192, OP = 576, kW = 3, kH = 3, dW = 1, dH = 1, padW = 1, padH = 1, iW = 12, iH = 12},
    {IP = 576, OP = 576, kW = 2, kH = 2, dW = 2, dH = 2, padW = 0, padH = 0, iW = 4, iH = 4}
}

torch.manualSeed(seed)
file = io.open("TorchPerform.csv", "a")
io.output(file)


for _, testCase in pairs(allTestCases) do

    local input = torch.Tensor(batchSize, testCase.IP, testCase.iW, testCase.iH)

    input:apply(function()
        i = torch.uniform(0, 1)
        return i
    end)

    local model = nn.SpatialConvolution(testCase.IP, testCase.OP,
        testCase.kW, testCase.kH, testCase.dW, testCase.dH, testCase.padW, testCase.padH)
    local timer = torch.Timer()
    for j = 1, forwardIterations do
        model:forward(input)
    end

    local timeSpend = timer:time().real / forwardIterations

    file:write("SpatialConvolution(" .. testCase.IP .. "," .. testCase.OP .. "," .. testCase.kW .."," ..
            testCase.kH .. "," .. testCase.dW .. "," .. testCase.dH .. "," .. testCase.padW .. "," .. testCase.padH .. ")forward:" .. string.format("%.6f", timeSpend) * 1000 .. "\n")
end

for _, testCase in pairs(allTestCases) do

    local input = torch.Tensor(batchSize, testCase.IP, testCase.iW, testCase.iH)
    input:apply(function()
        i = torch.uniform(0, 1)
        return i
    end
    )

    local model = nn.SpatialConvolution(testCase.IP, testCase.OP,
        testCase.kW, testCase.kH, testCase.dW, testCase.dH, testCase.padW, testCase.padH)
    local output = model:forward(input)
    local grad = output:clone():apply(function()
        i = torch.uniform(-2, 2)
        return i
    end)

    local timer = torch.Timer()
    for j = 1, backwardIterations do
        model:backward(input, grad)
    end

    local timeSpend = timer:time().real / backwardIterations
    file:write("SpatialConvolution(" .. testCase.IP .. "," .. testCase.OP .. "," .. testCase.kW .."," ..
            testCase.kH .. "," .. testCase.dW .. "," .. testCase.dH .."," .. testCase.padW .. "," .. testCase.padH .. ")backward:" .. string.format("%.6f", timeSpend) * 1000 .. "\n")
end

file:close()

File: dl/src/test/scala/com/intel/analytics/dllib/lib/nn/performTest/SpatialConvolutionPerform.scala
package com.intel.analytics.dllib.lib.nn.performTest

import com.intel.analytics.dllib.lib.nn._
import com.intel.analytics.dllib.lib.tensor.torch
import org.scalatest.{BeforeAndAfter, FlatSpec}

/**
  * Created by yao on 5/24/16.
  */
class SpatialConvolutionPerform extends FlatSpec with BeforeAndAfter{
  val ForwardIterations = PerformTestData.forwardIterations
  val BackwardIterations = PerformTestData.backwardIterations
  val Seed = 100
  val batchSize = 10

  val allTestCases = List(
    // AlexNet
    TestCase(3, 64, 11, 11, 4, 4, 2, 2 , 224, 224, 16, 35),
    TestCase(64, 192, 5, 5, 1, 1, 2, 2, 25, 25, 22, 55),
    TestCase(191, 384, 3, 3, 1, 1, 1, 1, 12, 12, 10, 35),
    TestCase(384, 256, 3, 3, 1, 1, 1, 1, 6, 6, 10, 25),
    TestCase(256, 256, 3, 3, 1, 1, 1, 1, 3, 3, 6, 10),
    //Cifar
    TestCase(3, 64, 3, 3, 1, 1, 1, 1, 224, 224, 70, 65),
    TestCase(64, 64, 3, 3, 1, 1, 1, 1, 110, 110, 90, 180),
    TestCase(64, 128, 3, 3, 1, 1, 1, 1, 54, 54, 40, 70),
    TestCase(128, 128, 3, 3, 1, 1, 1 ,1 ,26, 26, 19, 30),
    TestCase(128, 256, 3, 3, 1, 1, 1, 1, 13, 13, 7, 17),
    TestCase(256, 256, 3, 3, 1, 1, 1, 1, 6, 6, 7, 15),
    TestCase(256, 512, 3, 3, 1, 1, 1, 1, 3, 3, 7, 20),
    TestCase(512, 512, 3, 3, 1, 1, 1, 1, 2, 2, 12, 40),

    //GoogleNet
    TestCase(3, 64, 7, 7, 2, 2, 3, 3, 224, 224, 40, 55),
    TestCase(64, 64, 1, 1, 1, 1, 0, 0, 54, 54, 5, 8),
    TestCase(64, 192, 3, 3, 1, 1, 1, 1, 27, 27, 10, 25),
    TestCase(192, 576, 3, 3, 1, 1, 1, 1, 12, 12, 16, 42),
    TestCase(576, 576, 2, 2, 2, 2, 0, 0, 4, 4, 10, 25)
  )

  for (testCase <- allTestCases) {
    val input = torch.Tensor[Float](batchSize, testCase.nInputPlane, testCase.inputWidth, testCase.inputHeight).rand()
    val model = new SpatialConvolution[Float](testCase.nInputPlane, testCase.nOutputPlane,
      testCase.kW, testCase.kH, testCase.dW, testCase.dh, testCase.padW, testCase.padH)

    "A SpatialConvolutionPerform" should s" forward with a good performance with parameters" +
      s": ${testCase.nInputPlane}, ${testCase.nOutputPlane}, ${testCase.kW}, ${testCase.kH}, ${testCase.dW}, ${testCase.dh}," +
      s" ${testCase.padW}, ${testCase.padH}, ${testCase.inputHeight}, ${testCase.inputWidth}" in {
      if (!TestUtils.isRun())
        cancel("Performance regression tests are forbidden")
      val timeMillis = TestUtils.testModuleForwardPerform(input, model, ForwardIterations)
      assert(timeMillis < testCase.ForwardTimeoutMillis)
    }
  }

  for (testCase <- allTestCases) {
    val input = torch.Tensor[Float](batchSize, testCase.nInputPlane, testCase.inputWidth, testCase.inputHeight).rand()
    val model = new SpatialConvolution[Float](testCase.nInputPlane, testCase.nOutputPlane,
      testCase.kW, testCase.kH, testCase.dW, testCase.dh, testCase.padW, testCase.padH)
    val output = model.forward(input)
    val grad = output.clone().rand()

    "A SpatialConvolutionPerform" should s" backward with a good performance with parameters" +
      s": ${testCase.nInputPlane}, ${testCase.nOutputPlane}, ${testCase.kW}, ${testCase.kH}, ${testCase.dW}, ${testCase.dh}" +
      s", ${testCase.padW}, ${testCase.padH}, ${testCase.inputHeight}, ${testCase.inputWidth}" in {
      if (!TestUtils.isRun())
        cancel("Performance regression tests are forbidden")
      val timeMillis = TestUtils.testModuleBackwardPerform(input, grad, model, BackwardIterations)
      assert(timeMillis < testCase.BackwardTimeoutMillis)
    }
  }

  case class TestCase(nInputPlane : Int, nOutputPlane : Int, kW : Int, kH : Int, dW : Int, dh : Int,
                      padW: Int, padH: Int, inputWidth: Int, inputHeight: Int, ForwardTimeoutMillis: Int, BackwardTimeoutMillis: Int)
}

